crates/kornia-apriltag/src/threshold.rs

```diff
@@ -0,0 +1,198 @@
+use crate::{errors::AprilTagError, utils::PixelTrait};
+use kornia_image::{allocator::ImageAllocator, Image, ImageError};
+use std::ops::{Add, Div, Sub};
+
+/// TODO
+///
+/// ## Recommended values for `min_white_black_diff`
+///
+/// | Image Type | `min_white_black_diff` |
+/// |------------|------------------------|
+/// | 8-bit      | 10-20                  |
+/// | 16-bit     | 500-1000               |
+/// | 32-bit     | 100,000-1,000,000      |
+/// | float      | 0.05-0.1               |
+// TODO: Add support for parallelism
+pub fn adaptive_threshold<
+    T: PixelTrait
+        + PartialOrd
+        + Copy
+        + Add<T, Output = T>
+        + Sub<T, Output = T>
+        + Div<T, Output = T>
+        + From<u8>,
+    A1: ImageAllocator,
+    A2: ImageAllocator,
+>(
+    src: &Image<T, 1, A1>,
+    dst: &mut Image<T, 1, A2>,
+    tile_size: usize,
+    min_white_black_diff: T,
+) -> Result<(), AprilTagError> {
+    if src.size() != dst.size() {
+        return Err(
+            ImageError::InvalidImageSize(src.cols(), src.rows(), dst.cols(), dst.rows()).into(),
+        );
+    }
+
+    let src_data = src.as_slice();
+    let dst_data = dst.as_slice_mut();
+
+    let tiles_x_len = (src.width() as f32 / tile_size as f32).ceil() as usize; // number of horizontal tiles
+    let tiles_y_len = (src.height() as f32 / tile_size as f32).ceil() as usize; // number of vertical tiles
+
+    // pixels available in the tiles at edge
+    let last_tile_x_px = tile_size - (src.width() % tile_size);
+    let last_tile_y_px = tile_size - (src.height() % tile_size);
+
+    let mut tile_min: Vec<T> = Vec::with_capacity(tile_size * tile_size);
```

Comment:
Should we implement this as a stateful struct so that we don't have to allocate this buffers every time ?